lib/stdlib/src/lists.erl

```diff
@@ -1196,11 +1221,13 @@ enumerate_1(_Index, _Step, []) ->
 
 %%% Suggestion from OTP-2948: sort and merge with Fun.
 
--spec sort(Fun, List1) -> List2 when
-      Fun :: fun((A :: T, B :: T) -> boolean()),
-      List1 :: [T],
-      List2 :: [T],
-      T :: term().
+-spec sort(Fun, []) -> [] when
+        Fun :: fun((A :: T, B :: T) -> boolean());
+    (Fun, List1) -> List2 when
+        Fun :: fun((A :: T, B :: T) -> boolean()),
+        List1 :: [T,...],
+        List2 :: [T,...],
+        T :: term().
```

Comment:
Thanks for the PR! :)

I hate to bring this up since most of these specs were broken to begin with, but variables in type specs don't work the way most people would expect them to. Under the documented interpretation, these variables always refer to a _specific value_. That is while a variable `V` can have any value that satisfies all constraints, if repeated, it has the same _value_ everywhere. This has very funny implications such as the `Fun` here only returning normally when `A` and `B` are the same value.

This may seem crazy but is a natural consequence of how `dialyzer` works, and unfortunately the current type language is tightly coupled with how that tool sees things. We're working on finding a better model.

Can you update the specs you've touched so that they don't repeat variables?

Reply:
> these variables always refer to a _specific value_.

To a specific `value` or `type`?
E.g. is the following theoretically correct:
```erlang
-spec id(X) -> X when X :: integer().
id(0) -> 1.
```
?

Reply:
> To a specific value or type?

They are one and the same. `dialyzer` reasons purely about _sets of possible values_. There are no types in the common sense of the word.

>  E.g. is the following theoretically correct:

Nope.

Reply:
> They are one and the same. dialyzer reasons purely about sets of possible values. There are no types in the common sense of the word.

To elaborate, consider the call `id(0)`: the set of possible values for the argument is `0` alone. This conflicts with the return value `1`.

This would hold even if the body of the function was `id(_) -> 1.` It would only work for `id(1)`, because only then would `X` be the same in all positions.

Reply:
So, all specs in form of `[T,...]` are also wrong because that would mean non_empty list of the same value?

> Nope

Then why doesn't `dialyzer` emit an error here?

Reply:
> So, all specs in form of [T,...] are also wrong because that would mean non_empty list of the same value?

Lists are weird as they describe a recursive primitive so `[integer(),..]` will be interpreted as a non-empty list that contains any number of any integers, but yes, if `T` is repeated elsewhere (other than a list I guess? things are super-unclear here :-/) then the only interpretation that makes sense is that all cells have the same value.

> Then why doesn't dialyzer emit an error here?

Because it's woefully (and unevenly) lax when it comes to enforcing these equality constraints, and it's hard to make it stricter because it'd break so much code out there.

Reply:
So confused RN... So I used variables wrongly for a few years and I never got an error related to this. Do you have any example where this misusage  can cause errors? 

I used to think that variables in specs are alias like, so that the following two specs should be equal:
```erlang
spec f1(A :: C, B :: C) -> term() when C :: integer().

spec f2(A :: integer(), B :: integer()) -> term().
```
But yeah, changing anything here would break a lot of code...

You mentioned that you're working on a better model for this. At what stage is it? Is it a new tool, new `dialyzer` mode or something? I love static analysis so I'm curious about that stuff

Reply:
> Do you have any example where this misusage can cause errors?

More or less never.

> You mentioned that you're working on a better model for this. At what stage is it? Is it a new tool, new dialyzer mode or something? I love static analysis so I'm curious about that stuff

We're going to work a bit more on types in the mid-to-near future and it's in such early stages that I can't really say much about it, but as far as this topic is concerned it's probably just going to be removing the equality-constraint stuff from the documentation because it's super-confusing.